docs: add kerberos to nfs-ganesha design (#10732)
rook/rook: design/ceph/ceph-nfs-ganesha.md

          # limits:
          #   cpu: "2"
          #   memory: "1024Mi"

    kerberos:
      principalName: nfs
      configFiles:
        volumeSource:
          configMap:
            name: rook-ceph-nfs-organization-krb-conf

>       keytabFile:

Reviewer: `keytab` is the name, check `man 7 kerberos` for reference

>       keytabFile:

Author: Can definitely change to `keytabFile`, but I think `File` is worth keeping to help users understand it is the file they are providing and not the keytab text. This follows the same pattern as `sssdConfigFile` in the section above.

- `/run/dbus`: using the DBus instance from the sidecar caused SSSD errors in testing. SSSD only
  uses DBus for internal communications and creates its own socket as needed.

### Kerberos
Kerberos is the authentication mechanism natively supported by NFS-Ganesha.

#### Kerberos principals
The Kerberos service principal used by NFS-Ganesha to authenticate with the Kerberos server is built
up from 3 components:
1. a configured PrincipalName (from `spec.security.kerberos.principalName` in Rook) that acts as the
   service name
2. the hostname of the server on which NFS-Ganesha is running (using kernel call `getaddrinfo()`)
3. the realm (as configured by the user's krb.conf file(s))

The full service principal name is constructed as `<principalName>/<hostname>@<realm>`.

Users must add this service principal to their Kerberos server configuration. Therefore, this
principal must be static, and for the benefit of users it should be deterministic.

The hostname of Kubernetes pods is the partly-random name of the pod by default.

> In order to give NFS-Ganesha server pods a deterministic hostname, the `hostname` field of the pod

Reviewer 2: The `hostname` is set as an env var, or what field?

Author: Pods have a `hostname` field that can be set in the spec.

```
%url "rados://.nfs/${nfs-name}/kerberos"
```

These steps can be done from the Rook operator.

#### Kerberos config
Rook should take steps to remove any default configuration. This means that it should create its own
minimal krb.conf and ensure that any imported directories are empty. While it might be nice for some
users to include the default configurations that are present in the container, it is extremely
difficult in practice to adequately document the interactions between defaults that may change from
time to time in the container image (or between different distros), and users will have to expend
mental effort to understand how their configurations may override defaults. Upstream documentation
for krb5.conf is unclear about file ordering and override behavior. Therefore, Rook will rely on
users to specify nearly all of the configuration which will ensure users are able to easily supply
the configuration they require.

The minimal `/etc/krb5.conf` file Rook will create is as so:
```ini
includedir /etc/krb5.conf.d/ # include all user-defined config files

[logging]
default = STDERR # only log to stderr by default
```

#### Enabling kerberos for an export

> As of Ceph v17.2.3, the `ceph nfs ...` CLI tool is unable to create exports with Kerberos security

Reviewer 2: Is there a known fix, or unknown at this point? How about linking to a ceph issue? Unless we can know the version where it's fixed, I'm not sure it's helpful to mention the currently version.
```suggestion
Currently the `ceph nfs ...` CLI tool is unable to create exports with Kerberos security
```

Author: When the fix will be available is unknown.

- `/run/dbus`: using the DBus instance from the sidecar caused SSSD errors in testing. SSSD only
  uses DBus for internal communications and creates its own socket as needed.

### Kerberos
**NOTE:** The principles behind this design have not been validated.

#### Volumes
Volumes that should be mounted into nfs-ganesha container to support Kerberos:
1. `keytabFile` volume: use `subPath` on the mount to add the `krb5.keytab` file to `/etc/krb5.keytab`
2. `configFiles` volume: mount (without `subPath`) to `/etc/krb5.conf.rook/` to allow all files to
   be mounted (e.g., if a ConfigMap has multiple data items or hostPath has multiple conf.d files)

#### Ganesha config
Should add configuration. Docs say Active_krb 5 is default true if krb support is compiled in, but
most examples have this explicitly set.
Default PrincipalName is "nfs".

> Default for keytab path is reportedly empty. Rook can use `/etc/ker

Reviewer 2: is this the expected path or typo?
```suggestion
Default for keytab path is reportedly empty. Rook can use `/etc/ker`
```

Reviewer: I assume a typo, `/etc/krb5.keytab` is the default

Create a new RADOS object named `kerberos` to configure Kerberos.
```ini
NFS_KRB5
{

>    PrincipalName = nfs ; # <-- set from spec.security.kerberos.principalName

Author: Note from prototyping: double quotes seem to be required here.

```suggestion
   PrincipalName = "nfs" ; # <-- set from spec.security.kerberos.principalName
```

    kdc = FILE:/var/log/krb5kdc.log
    admin_server = FILE:/var/log/kadmind.log

[libdefaults]
    dns_lookup_realm = false
    ticket_lifetime = 24h
    renew_lifetime = 7d
    forwardable = true
    rdns = false
    pkinit_anchors = FILE:/etc/pki/tls/certs/ca-bundle.crt
    spake_preauth_groups = edwards25519
#    default_realm = EXAMPLE.COM
    default_ccache_name = KEYRING:persistent:%{uid}

[realms]
# EXAMPLE.COM = {
#     kdc = kerberos.example.com
#     admin_server = kerberos.example.com
# }

[domain_realm]
# .example.com = EXAMPLE.COM
# example.com = EXAMPLE.COM
```

Rook should set its own krb.conf like so. Rook can create `krb5.conf` in an EmptyDir from an init
container and mount it to the `nfs-ganesha` container in the same fashion as for `nsswitch.conf`.
Inline comments explain changes made compared to the default file.
```ini
includedir /etc/krb5.conf.d/    # assume we want to keep crypto-policies config
includedir /etc/krb5.conf.rook/ # load configFiles content from users

[logging]
    default = STDERR # only log to stderr by default

# remove all other defaults (namely 'libdefaults') so that users have full flexibility to set what they want.
```

> **QUESTION:** Do we want to keep this `crypto-policies` file? Do we think users might want to remove

Reviewer 2: You're still prototyping, right? Do you expect these questions to be answered with a prototype?

it? Will overriding `[libdefaults] permitted_enctypes` be sufficient? Will they have to name their

> file `zz-*` to ensure it overwrites `crypto-policies`?

Reviewer: I'd recommend keeping the files that come with the distributions packaging. They are there probably for a reason. If users have the ability to override the standard settings, that should be sufficient IMHO.

```
$# cat /etc/krb5.conf.d/crypto-policies
[libdefaults]
permitted_enctypes = aes256-cts-hmac-sha1-96 aes256-cts-hmac-sha384-192 camellia256-cts-cmac aes128-cts-hmac-sha1-96 aes128-cts-hmac-sha256-128 camellia128-cts-cmac
```

#### Enabling kerberos for an export
**QUESTION:**
`sectype` enables Kerberos for an export and is set on the Export definition. I don't see that
option in `ceph nfs export ...` commands. Do we need to manually add `sectype` to all exports?

> How will CSI do this? Should be part of Ceph API, no?

Reviewer: It should get part of the Ceph-API, yes. Possibly other options need to get included there as well.

Author: Is this something you could start working on? I wonder if it would be good to have the ability to add generic export options rather than continuously adding flags to the Ceph CLI API when we find something is missing.
